### PR TITLE
Vote for aggregated public key

### LINKED ISSUE #3968: [Clarity] DKG vote registration
Signers should be able to submit their DKG vote to the clarity smart contract

May be able to reuse vote DKG registration code within https://github.com/Trust-Machines/stacks-sbtc/blob/main/sbtc-mini/contracts/sbtc-registry.clar

### PR DESCRIPTION
**Pull request by friedger**
### Description
This PR adds voting for the aggregated public key submitted by signers.

### Applicable issues

- fixes #3968 

### Checklist

- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

File: contrib/core-contract-tests/contracts/pox/pox-4-aggregated-public-key.clar [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+(define-read-only (get-signer-public-key (signer principal) (reward-cycle uint))
+    (some 0x0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20))
+
+(define-read-only (get-signer-weight (signer-public-key (buff 33)) (reward-cycle uint))
+    u1000000000000)
+
+(define-read-only (current-reward-cycle)
+    u0)
+
+(define-public (vote-for-aggreated-public-key (key (buff 33)) (reward-cycle uint) (round uint) (tapleaves (list 4001 (buff 33))))
```
Comment (friedger) [inline contrib/core-contract-tests/contracts/pox/pox-4-aggregated-public-key.clar:31]: @setzeus What must be contained in a vote?

File: contrib/core-contract-tests/contracts/pox/pox-4-aggregated-public-key.clar [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    ;; (contract-call? .pox-4 get-signer-public-key reward-cycle signer)
+    (some 0x0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20))
+
+(define-read-only (get-signer-slots (signer-public-key (buff 33)) (reward-cycle uint))
+    u1000000000000)
+
+(define-read-only (current-reward-cycle)
+    u0)
+
+(define-public (vote-for-aggreated-public-key (key (buff 33)) (reward-cycle uint) (round uint) (tapleaves (list 4001 (buff 33))))
```
Comment (jferrant) [inline contrib/core-contract-tests/contracts/pox/pox-4-aggregated-public-key.clar:42]: Typo "aggreated"

File: stackslib/src/chainstate/stacks/boot/pox-4.clar [hunk 3]
```diff
@@ -1361,4 +1361,61 @@
...
+;;
+;; @contract voting for the aggregated public key
+;;
+
+;; maps dkg round and signer to proposed aggregated public key
+(define-map votes {reward-cycle: uint, round: uint, signer: principal} {aggregated-public-key: (buff 33), reward-slots: uint})
+;; maps dkg rount and aggregated public key to weights of signers supporting this key so far
+(define-map tally {reward-cycle: uint, round: uint, aggregate-public-key: (buff 33)} uint)
+
+(define-constant err-not-allowed (err u10000))
```
Comment (setzeus) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:1375]: Can you relocate to batch w/ rest of errors & update the format to match existing? (capitalized, currently in the 30s index).

`(define-constant ERR_INCORRECT_REWARD_CYCLE 32)`

File: stackslib/src/chainstate/stacks/boot/pox-4.clar [hunk 4]
```diff
@@ -1361,4 +1361,61 @@
...
+    ;; (contract-call? .pox-4 get-signer-public-key reward-cycle signer)
+    (some 0x0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20))
+
+(define-read-only (get-signer-slots (signer-public-key (buff 33)) (reward-cycle uint))
+    u1000000000000)
+
+(define-read-only (current-reward-cycle)
+    u0)
+
+(define-public (vote-for-aggregated-public-key (key (buff 33)) (reward-cycle uint) (round uint) (tapleaves (list 4001 (buff 33))))
```
[the remark this answers is not in the document]
Comment (friedger) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:1409]: I don't know where to get the weight of a stacker from. Should I rebase on https://github.com/stacks-network/stacks-core/pull/4164 ?

**Comment by friedger** [comment]
Replaced by #4239

**Pull request closed without merging**